### ISSUE TITLE
Scaling Horovod using CIFAR

### ISSUE THREAD

**Issue by abidmalikwaterloo** [horovod/horovod#1273]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
Pytorch
2. Framework version:
1
3. Horovod version:
latest
4. MPI version:
5. CUDA version:
9
6. NCCL version:
2
7. Python version:
3
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.

Time per epoch for single GPU is less than 4 GPUs. It does not make sense.

**Comment by alsrgv** [comment]
@abidmalikwaterloo, what's your hardware setup? Is your CIFAR model training script published anywhere?

**Comment by abidmalikwaterloo** [comment]
I am using Summit machine. Each node has 6 V100 GPUs. The model training script is very simple. I used horovod/example template

```
def main():
    import horovod.torch as hvd
    import sys
    hvd.init()

    global args, best_prec1
    args = parser.parse_args()

    if (torch.cuda.is_available()):
        torch.cuda.set_device(hvd.local_rank())
        torch.cuda.manual_seed(42)



    kwargs = {'num_workers':1, 'pin_memory':True} if torch.cuda.is_available() else {}


    traindir = os.path.join(args.data, 'train')
    valdir = os.path.join(args.data, 'val')

    print('train image path : ', str(traindir))
    print('validation image path : ', str(valdir))

    transform_train = transforms.Compose([transforms.ToTensor(),transforms.Normalize((0.4914,0.4822,0.4465),(0.2023,0.1994,0.2010)),])
    transform_test  = transforms.Compose([transforms.ToTensor(),transforms.Normalize((0.4914,0.4822,0.4465),(0.2023,0.1994,0.2010)),])

    import torchvision

    trainset = torchvision.datasets.CIFAR10( root=traindir, train=True, download=True, transform=transform_train)
    train_sampler = torch.utils.data.distributed.DistributedSampler( trainset, num_replicas=hvd.size(), rank=hvd.rank())
    trainloader = torch.utils.data.DataLoader( trainset, batch_size=args.batch_size, sampler=train_sampler, num_workers=0)



    testset = torchvision.datasets.CIFAR10( root=valdir, train=False, download=True, transform=transform_test)
    test_sampler = torch.utils.data.distributed.DistributedSampler( testset, num_replicas=hvd.size(), rank=hvd.rank())
    testloader = torch.utils.data.DataLoader( testset, batch_size=args.batch_size, sampler=test_sampler, num_workers=0)

    global model
    model = models.vgg16()

if torch.cuda.is_available():
        model.cuda()

    import torch.optim as optim
    import torch.nn.functional as F

    criterion = nn.CrossEntropyLoss().cuda()
    optimizer = optim.SGD(model.parameters(), lr=0.0001 , momentum=args.momentum)
    hvd.broadcast_parameters(model.state_dict(), root_rank=0)
    hvd.broadcast_optimizer_state(optimizer, root_rank=0)
    compression = hvd.Compression.fp16 if args.fp16_allreduce else hvd.Compression.none
    optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters(), compression=compression)

    #mytrain()
#def mytrain():
    print("I am in the training phase")
    sys.stdout.flush()
    model.train()
    train_sampler.set_epoch(10)
    print("starting the loop")
    print("The length of the batch", len(trainloader))
    for epoch in range(0, 10):
        start=time.time()
        for batch_idx, (data, target) in enumerate(trainloader):
            if torch.cuda.is_available():
                data = data.cuda()
                target = target.cuda()
            optimizer.zero_grad()
            output = model(data)
            loss = criterion(output, target)
            loss.backward()
            optimizer.step()
          #  print(epoch)
        end=time.time()
        print( loss.item())
        sys.stdout.flush()
 sys.stdout.flush()
        print(" I am done with epoch", epoch)
        sys.stdout.flush()
        print(" The training time is ", (end - start))





if __name__ == '__main__':


    #dist.init_process_group('mpi')

    main()

**Comment by abidmalikwaterloo** [comment]
@alsrgv I just tried the implementation without Horovod. The running time per epoch with CIFAR using one GPU is better than multi GPUs with batch size 32 (data points) using Horovod.

**Comment by abditag2** [comment]
@abidmalikwaterloo In your current setup, if you use N GPUs, you get a effective batch size of N*32. 
You can reduce the batch size of each worker to `batch_size=args.batch_size/hvd.size()` and then compare completion of one epoch in single GPU and multi  GPU cases. 

Also, in order to get the best results, you should scale your learning rate with the number of workers i.e., `learning_rate = arg.lr * hvd.size()`

**Comment by abidmalikwaterloo** [comment]
@abditag2 Does Horovod benefit from cuda aware MPI?

**Comment by abditag2** [comment]
what do you mean by CUDA aware MPI? MPI and Gloo are mainly used in the control layer of Horovod (i.e., to do all the coordination between the workers). They can also perform the data layer operations for gradient aggregation however you will get much better results with NCCL which uses CUDA and GPUs under the hood.

**Comment by karakusc** [comment]
@abidmalikwaterloo If MPI is installed with CUDA support, and if Horovod is built with `HOROVOD_GPU_ALLREDUCE=MPI` or `HOROVOD_GPU_ALLGATHER=MPI`, Horovod can leverage this by performing the MPI collectives directly from the GPU memory. Whether this will have a benefit will likely depend on the system and the model you are training, I believe. In most cases, you will get better results with NCCL as @abditag2 said, or something hybrid like hierarchical allreduce.
